aioxmpp/pep/service.py

```diff
@@ -0,0 +1,238 @@
+
+        Dispatch event notifications for `node_namespace` to `handler`.
+
+        This registers `node_namespace` as feature for service discovery
+        unless ``register_feature=False`` is passed.
+
+        :param node_namespace: the pubsub node whose events shall be
+            handled.
+        :param handler: the handler to install.
+        :type handler: callable, see
+            :attr:`aioxmpp.PubSubClient.on_item_publish`
+            for the arguments.
+        :param register_feature: Whether to publish the `node_namespace`
+            as feature.
+        :param notify: Whether to register the ``+notify`` feature to
+            receive notification without explicit subscription.
+
+        :raises RuntimeError: if a handler for `node_namespace` is already
+            set.
+        """
+        if node_namespace in self._pep_node_claims:
+            raise RuntimeError(
+                "setting handler for already handled namespace"
+            )
+        if register_feature:
+            self._disco_server.register_feature(node_namespace)
+        if notify:
+            self._disco_server.register_feature(node_namespace + "+notify")
+        self._pep_node_claims[node_namespace] = handler, register_feature, notify
+
+    def unclaim_pep_node(self, node_namespace):
+        """
+        Unclaim `node_namespace`.
+
+        The feature for `node_namespace` and the ``+notify`` feature
+        are automatically retracted if they were set by
+        :method:`claim_pep_node`.
+
+        :param node_namespace: The PubSub node whose handler shall be unset.
+
+        :raises KeyError: If the no handler is registered for
+            `node_namespace`.
+        """
+        _, feature, notify = self._pep_node_claims.pop(node_namespace)
+        if notify:
+            self._disco_server.unregister_feature(node_namespace + "+notify")
+        if feature:
+            self._disco_server.unregister_feature(node_namespace)
+
+    @asyncio.coroutine
+    def _check_for_pep(self):
+        # XXX: should this be done when the stream connects
+        # and we use the cached result later on (i.e. disable
+        # the PEP service if the server does not support PEP)
+        disco_info = yield from self._disco_client.query_info(
+            self.client.local_jid.bare()
+        )
+
+        for item in disco_info.identities.filter(attrs={"category": "pubsub"}):
+            if item.type_ == "pep":
+                break
+        else:
+            raise RuntimeError("server does not support PEP")
+
+    @service.depsignal(aioxmpp.PubSubClient, "on_item_published")
+    def _handle_pubsub_publish(self, jid, node, item, *, message=None):
+        try:
+            handler, _, _ = self._pep_node_claims[node]
+        except KeyError:
+            return
+
+        # TODO: handle empty payloads due to (mis-)configuration of
+        # the node specially.
+        handler(jid, node, item, message=message)
```

Comment:
If we don’t care about synchronicity, you could pass ``defer=True`` to ``depsignal`` and make ``_handle_pubsub_publish`` a coroutine. That would allow to retrieve the payload if it isn’t sent.